aten/src/ATen/native/mkldnn/TensorShape.cpp

```diff
@@ -38,9 +42,8 @@ Tensor mkldnn_view(const Tensor& self, IntArrayRef size) {
 Tensor mkldnn_reshape(const Tensor& self, IntArrayRef size) {
   auto inferred_size = at::infer_size(size, self.numel());
   const ideep::tensor& x = itensor_from_mkldnn(self);
-  ideep::tensor y;
-  ideep::direct_copy::compute<AllocForMKLDNN>(x, y);
-  y.reshape({inferred_size.cbegin(), inferred_size.cend()});
+  ideep::tensor y{x};
+  y.reshape<AllocForMKLDNN>({inferred_size.cbegin(), inferred_size.cend()});
```

Comment:
Hmm if new shapes are as same as the old shapes, [`ideep::tensor::reshape`](https://bddppq.github.io/codebrowser/pytorch/pytorch/third_party/ideep/include/ideep/tensor.hpp.html#_ZN5ideep6tensor7reshapeERKSt6vectorIiSaIiEE) will not create a copy and thus y will share the same memory as x. This could be messy since then there will be two aten opaque tensors sharing the same underlying ideep tensor.

Reply:
@bddppq We wanted to keep the semantics of `reshape` as much as possible here, i.e. do shallow copy on plain-format MKL-DNN tensor (layout of which is contiguous) and do deep copy otherwise (see https://github.com/intel/ideep/blob/d7304e0345c3f0647cb020a71682d680b9f0424a/include/ideep/tensor.hpp#L1100). For the shallow copy case, `ideep::tensor y{x}` would make `y` copy the metadata of `x` while share its underlying storage. This is the same semantics as the shallow copy of a native CPU tensor. So this should work fine here.

Reply:
It's fine to follow the semantics of cpu tensor to return the input tensor as output if the shapes are the same, but it's not fine to let two different aten opaque tensors to hold the same underlying ideep tensor.

Reply:
@bddppq But `y` and `x` are two different `ideep::tensor` object, right? They just share the same underlying storage while hold their own metadata. It is similar to how two CPU tensors share the same underlying storage while have their own metatdata. Make sense?

Reply:
@bddppq @XiaobingSuper OK. Sounds like we should call `y.set_descriptor(x.get_descriptor())` explicitly after `ideep::tensor y{x}` in order to guarantee `y` uses a new MKL-DNN descriptor while sharing the underlying buffer of `x`. We should have wrapped an explicit shallow copy API in ideep to do such a thing actually.

Reply:
@jgong5 @XiaobingSuper I would say just add the following line at the beginning of this function :-)
```c++
if (self.sizes() == inferred_size) {
  return self;
}
```

Reply:
Yes, I will change it. Thanks!

Reply:
@bddppq, @jgong5 , changed it, thanks!

Reply:
@bddppq `y.reshape` would create a new descriptor by calling `set_descriptor()` so additional `y.set_descriptor(x.get_descriptor())` is not necessary here. With the recent change of @XiaobingSuper , the behavior of `mkldnn_reshape` looks like below, mimicking the semantics of `reshape` of native CPU tensor:
1. If the requested shape is the same as `self`, return `self` directly.
2. If the format of MKL-DNN tensor is public (or plain), the storage is contiguous, the returned MKL-DNN tensor shares the storage of `self` and copies all metadata of `self` with the new shape.
3. For other situations, return a deep-copied tensor of `self` with the new shape.

Reply:
@jgong5 @XiaobingSuper This sounds good to me.